src/languages/latex.js

```diff
@@ -1,70 +1,263 @@
+      '(?:(?:e|g|x)?def|let)',
+      '(?:begin|end)',
+      '(?:part|chapter|(?:sub){0,2}section|(?:sub)?paragraph)',
+      'caption',
+      '(?:label|(?:eq|page|name)?ref|(?:paren|foot|super)?cite)'
+    ].map(word => word + '(?![a-zA-Z@:_])'));
+  const KNOWN_CONTROL_SYMBOLS = /\\|\(|\)|\[|\]|\s|!|,|;/
+  const L3_REGEX = new RegExp([
+      '(?:__)?[a-zA-Z]{2,}_[a-zA-Z](?:_?[a-zA-Z])+:[a-zA-Z]*',
+      '[lgc]__?[a-zA-Z]+_[a-zA-Z](?:_?[a-zA-Z])*[a-zA-Z]',
+      '[qs]__?[a-zA-Z](?:_?[a-zA-Z])+',
+      'use(?:_i)?:[a-zA-Z]*',
+      '(?:else|fi|or):',
+      '(?:if|cs|exp):w',
+      '(?:hbox|vbox):n',
+      '::[a-zA-Z]_unbraced',
+      '::[a-zA-Z:]'
+    ].map(pattern => pattern + '(?![a-zA-Z:_])').join('|'));
+  const L2_VARIANTS = [
+    {begin: /[a-zA-Z@]+/}, // control word
+    {begin: /[^a-zA-Z@]?/} // control symbol
+  ];
+  const DOUBLE_CARET_VARIANTS = [
+    {begin: /\^{6}[0-9a-f]{6}/},
+    {begin: /\^{5}[0-9a-f]{5}/},
+    {begin: /\^{4}[0-9a-f]{4}/},
+    {begin: /\^{3}[0-9a-f]{3}/},
+    {begin: /\^{2}[0-9a-f]{2}/},
+    {begin: /\^{2}[\u0000-\u007f]/}
+  ];
+  const CONTROL_SEQUENCE = {
+    className: 'keyword',
     begin: /\\/,
-    relevance: 0,
     contains: [
       {
-        className: 'name',
+        endsParent: true,
         variants: [
-          {begin: /[a-zA-Z\u0430-\u044f\u0410-\u042f]+[*]?/},
-          {begin: /[^a-zA-Z\u0430-\u044f\u0410-\u042f0-9]/}
-        ],
-        starts: {
-          endsWithParent: true,
-          relevance: 0,
-          contains: [
-            {
-              className: 'string', // because it looks like attributes in HTML tags
-              variants: [
-                {begin: /\[/, end: /\]/},
-                {begin: /\{/, end: /\}/}
-              ]
-            },
-            {
-              begin: /\s*=\s*/, endsWithParent: true,
-              relevance: 0,
-              contains: [
-                {
-                  className: 'number',
-                  begin: /-?\d*\.?\d+(pt|pc|mm|cm|in|dd|cc|ex|em)?/
-                }
-              ]
-            }
-          ]
-        }
+          {begin: KNOWN_CONTROL_WORDS},
+          {begin: KNOWN_CONTROL_SYMBOLS}
+        ]
+      },
+      {
+        endsParent: true,
+        begin: L3_REGEX
+      },
+      {
+        endsParent: true,
+        variants: DOUBLE_CARET_VARIANTS
+      },
+      {
+        endsParent: true,
+        relevance: 0,
+        variants: L2_VARIANTS
       }
     ]
   };
-
-  return {
-    name: 'LaTeX',
-    aliases: ['tex'],
-    contains: [
-      COMMAND,
+  const MACRO_PARAM = {
+    className: 'params',
+    relevance: 0,
+    begin: /#+\d?/
+  };
+  const DOUBLE_CARET_CHAR = {
+    // relevance: 1
+    variants: DOUBLE_CARET_VARIANTS
+  };
+  const SPECIAL_CATCODE = {
+    className: 'built_in',
+    relevance: 0,
+    begin: /[$&^_]/
+  };
+  const BRACES = {
+    className: 'built_in',
+    relevance: 0,
+    begin: /[{}]/
+  };
+  const MAGIC_COMMENT = hljs.COMMENT(
```

Comment:
Would this typically have "normal" comment explanation text or doctags?  All the examples I saw looked VERY specific, not general purpose comments.

If so we probably shouldn't use the COMMENT mode but rather just turn this into a normal rule. (to avoid all the extra junk COMMENT carries with it which I think may not apply here.) Thoughts?

Reply:
No, there would not be any normal text in a magic comment. I agree that we thus shouldn't use `COMMENT`.